### PR TITLE
nn.NarrowTable

### PR DESCRIPTION
**Pull request by nicholas-leonard**
includes unit tests, doc.

### CODE REVIEW

File: NarrowTable.lua [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+   if not offset then
+      error('nn.NarrowTable(offset, length)')
+   end
+   
+   self.output = {}
+   self.gradInput = {}
+end
+
+function NarrowTable:updateOutput(input)
+   for k,v in ipairs(self.output) do self.output[k] = v end
```
Comment (soumith) [inline NarrowTable.lua]: wait, what is going on here? is this a redundant line?

File: NarrowTable.lua [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+   for i=1,#input do
+      if (i < self.offset) or (i >= self.offset + self.length) then
+         self.gradInput[i] = nn.utils.recursiveResizeAs(self.gradInput[i], input[i])
+         nn.utils.recursiveFill(self.gradInput[i], 0)
+      end
+   end
+   return self.gradInput
+end 
+
+function NarrowTable:type(type)
```
Comment (soumith) [inline NarrowTable.lua]: There's a PR we are thinking of merging soon, it requires a small change here: https://github.com/torch/nn/pull/303/files

Can you do the change here as well.

function NarrowTable:type(type, tensorCache)

File: NarrowTable.lua [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+   self.output = {}
+   self.gradInput = {}
+   return parent.type(self, type)
```
Comment (soumith) [inline NarrowTable.lua]: return parent.type(self, type, tensorCache)

File: NarrowTable.lua (hunk 1 above)
Comment (nicholas-leonard) [inline NarrowTable.lua]: No it just empties the table. That ways self.output always points to the same table.

Comment (soumith) [inline NarrowTable.lua]: hmm, I am very confused. You are just reassigning all the integer keys of self.output to their values. I dont see what has changed after this line, self.output is exactly the same.

Comment (nicholas-leonard) [inline NarrowTable.lua]: Hmm yes this is indeed an error. should be = nil.

Nicholas Léonard
917-741-7570

**Comment by nicholas-leonard** [comment]
Ok. changes made. Added unit test to test variable length input.

**Comment by nicholas-leonard** [comment]
This going in?

**Comment by soumith** [comment]
Yes, for sure

**Comment by soumith** [comment]
thanks!

**Pull request merged**